protofsm/state_machine_test.go

```diff
@@ -0,0 +1,330 @@
+		"02ec95e4e8ad994861b95fc5986eedaac24739e5ea3d0634db4c8ccd44cd" +
+			"a126ea",
+	))
+	pub2, _ = btcec.ParsePubKey(hexDecode(
+		"0356167ba3e54ac542e86e906d4186aba9ca0b9df45001c62b753d33fe06" +
+			"f5b4e8",
+	))
+)
+
+func (d *dummyStateStart) ProcessEvent(event dummyEvents, env *dummyEnv,
+) (*StateTransition[dummyEvents, *dummyEnv], error) {
+
+	switch event.(type) {
+	case *goToFin:
+		return &StateTransition[dummyEvents, *dummyEnv]{
+			NextState: &dummyStateFin{},
+		}, nil
+
+	// This state will loop back upon itself, but will also emit an event
+	// to head to the terminal state.
+	case *emitInternal:
+		return &StateTransition[dummyEvents, *dummyEnv]{
+			NextState: &dummyStateStart{},
+			NewEvents: fn.Some(EmittedEvent[dummyEvents]{
+				InternalEvent: fn.Some(dummyEvents(&goToFin{})),
+			}),
+		}, nil
+
+	// This state will proceed to the terminal state, but will emit all the
+	// possible daemon events.
+	case *daemonEvents:
+		// This send event can only succeed once the bool turns to
+		// true. After that, then we'll expect another event to take us
+		// to the final state.
+		sendEvent := &SendMsgEvent[dummyEvents]{
+			TargetPeer: *pub1,
+			SendWhen: fn.Some(func() bool {
+				return d.canSend.Load()
+			}),
+			PostSendEvent: fn.Some(dummyEvents(&goToFin{})),
+		}
+
+		// We'll also send out a normal send event that doesn't have
+		// any preconditions.
+		sendEvent2 := &SendMsgEvent[dummyEvents]{
+			TargetPeer: *pub2,
+		}
+
+		return &StateTransition[dummyEvents, *dummyEnv]{
+			// We'll state in this state until the send succeeds
+			// based on our predicate. Then it'll transition to the
+			// final state.
+			NextState: &dummyStateStart{
+				canSend: d.canSend,
+			},
+			NewEvents: fn.Some(EmittedEvent[dummyEvents]{
+				ExternalEvents: fn.Some(DaemonEventSet{
+					sendEvent, &DisableChannelEvent{},
+					&BroadcastTxn{}, sendEvent2,
+				}),
+			}),
+		}, nil
+	}
+
+	return nil, fmt.Errorf("unknown event: %T", event)
+}
+
+func (d *dummyStateStart) IsTerminal() bool {
+	return false
+}
+
+type dummyStateFin struct {
+}
+
+func (d *dummyStateFin) ProcessEvent(event dummyEvents, env *dummyEnv,
+) (*StateTransition[dummyEvents, *dummyEnv], error) {
+
+	return &StateTransition[dummyEvents, *dummyEnv]{
+		NextState: &dummyStateFin{},
+	}, nil
+}
+
+func (d *dummyStateFin) IsTerminal() bool {
+	return true
+}
+
+func assertState[Event any, Env Environment](t *testing.T,
+	m *StateMachine[Event, Env], expectedState State[Event, Env]) {
+
+	state, err := m.CurrentState()
+	require.NoError(t, err)
+	require.IsType(t, expectedState, state)
+}
+
+func assertStateTransitions[Event any, Env Environment](
+	t *testing.T, stateSub StateSubscriber[Event, Env],
+	expectedStates []State[Event, Env]) {
+
+	for _, expectedState := range expectedStates {
+		newState := <-stateSub.NewItemCreated.ChanOut()
+
+		require.IsType(t, expectedState, newState)
+	}
+}
+
+type dummyAdapters struct {
+	mock.Mock
```

Comment:
Niiice. I really like mocks as it makes testing and maintaining the code easier, plus it gives a stronger guarantee about the robustness of the code since you can also assert all aspects of a mocked method - how many times it's called, what are the params used, etc. Another benefit I find very useful to myself is it helps with understanding the workflow of a tested method, and can give feedback which leads to a better structure. That being said, I think if we mock all the interfaces here, we can even make the control flow more clear,

Reply:
Yep agreed on all counts, and then it forces you to create/accept interfaces in most areas (instead of say closures or w/e), which makes it all easier to test. More on that in the final PR in this series.